sdk/core/azure-core/test/ut/transport_adapter_curl.cpp

```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
+
+    // Make sure there are nothing in the pool
+    EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 0);
+
+    // Use the same request for all connections.
+    Azure::Core::Http::Request req(
+        Azure::Core::Http::HttpMethod::Get, Azure::Core::Http::Url("http://httpbin.org/get"));
+
     {
-      Azure::Core::Http::Url host("http://httpbin.org/get");
-      Azure::Core::Http::CurlConnectionPool::ClearIndex();
-
-      auto threadRoutine = [&]() {
-        using namespace std::chrono_literals;
-        auto request = Azure::Core::Http::Request(Azure::Core::Http::HttpMethod::Get, host);
-        auto response = m_pipeline->Send(Azure::Core::GetApplicationContext(), request);
-        checkResponseCode(response->GetStatusCode());
-        auto expectedResponseBodySize = std::stoull(response->GetHeaders().at("content-length"));
-        CheckBodyFromBuffer(*response, expectedResponseBodySize);
-        std::this_thread::sleep_for(1s);
-      };
-
-      std::thread t1(threadRoutine);
-      std::thread t2(threadRoutine);
-      t1.join();
-      t2.join();
-
-      // 2 connections must be available at this point
-      EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 2);
-
-      std::thread t3(threadRoutine);
-      std::thread t4(threadRoutine);
-      std::thread t5(threadRoutine);
-      t3.join();
-      t4.join();
-      t5.join();
-
-      // Two connections re-used plus one connection created
-      EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 3);
+      // Creating a new connection with default options
+      Azure::Core::Http::CurlTransportOptions options;
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          req,
+          Azure::Core::Http::CurlConnectionPool::GetCurlConnection(req, options),
+          options.HttpKeepAlive);
+      // Simulate connection was used already
+      session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+      session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
+    }
+    // Check that after the connection is gone, it is moved back to the pool
+    EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 1);
+
+    // Test that asking a connection with same config will re-use the same connection
+    {
+      // Creating a new connection with default options
+      Azure::Core::Http::CurlTransportOptions options;
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          req,
+          Azure::Core::Http::CurlConnectionPool::GetCurlConnection(req, options),
+          options.HttpKeepAlive);
+      // Simulate connection was used already
+      session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+      session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
+      EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 0);
+    }
```

Comment:
I am trying to understand how this is testing what the comment is suggesting, but haven't difficulty. How are we ensuring the same connection is being used exactly?

Reply:
Before creating the connection we check the number of connections in the pool, which is 1.
Then we create a new session and ask for a new connection to the pool.
Then we check the pool and validate that there are no connections in there. Which means the connection was re-used.